### PR TITLE
Refs #12716 - Extensible show page for smart proxy - rejected version

### PR DESCRIPTION
**Pull request by xprazak2**
This allows us to create extensible pages. Let me know if this is a step in the right direction.

We can register new extensible page with `add_page`. We can extend an already registered page (add more content to it) with `extend_page`. We can also extend the existing tabs with `extend_tab`. 
Several columns for page/tab can be specified and their content arranged on a page accordingly. Plugins can easily extend registered pages from within their `Engine`, there are usage examples in  services/foreman/plugin.rb

### CODE REVIEW

File: app/controllers/smart_proxies_controller.rb [hunk 1]
```diff
@@ -27,6 +24,17 @@ def edit
     @proxy = @smart_proxy
   end
 
+  def show
+    render_extensible
+  end
+
+  def ping
```
Comment (ohadlevy) [inline app/controllers/smart_proxies_controller.rb]: rebase error?

**Comment by lzap** [comment]
Nicely done, extensible pages is something we need a lot for years now.

Before I start with review and testing, I would like to discuss "widget" term. Since we already use this term for user's widgets, I suggest to name those it differently. It could be "snippet" (we already use this term as well), "partial" (but thats Rails one) or "portlet" (seems to be available). Any suggestions or you prefer to stay with "widgets"?

File: app/services/pages/view_item.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      yield tab if block_given?
+      @tabs[opts[:name]] = tab
+    end
+
+    def add_widget(opts)
+      @columns[opts[:column] ? opts[:column] : 0] << Pages::Widget.new(opts)
+    end
+
+    def sorted_tabs
+      @tabs.sort_by { |k,v| v.priority }.reverse.to_h
```
Comment (lzap) [inline app/services/pages/view_item.rb]: Getting error here: `undefined method`to_h' for #Array:0x0055dfbe6d6c50`

File: app/controllers/smart_proxies_controller.rb (hunk 1 above)
Comment (xprazak2) [inline app/controllers/smart_proxies_controller.rb]: Yep, `ping` method was already there and git marked it as a newly added...

**Comment by xprazak2** [comment]
@lzap, I used "widget" for the lack of better term, It can be definitely changed if we find a name that is less overloaded.

**Comment by lzap** [comment]
Can you edit the description with little more explanation of the concept here? I struggle a bit in recognizing `add_page` vs `extend_page`. Just few words.

File: app/controllers/smart_proxies_controller.rb (hunk 1 above)
Comment (ohadlevy) [inline app/controllers/smart_proxies_controller.rb]: ok, so i assume remove it?

File: app/services/pages/loader.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+module Pages
+  class Loader
+    def self.load
+      Pages::Manager.add_page({ :controller => :smart_proxies, :action => :show }, "smart_proxies/show") do |page|
```
Comment (ohadlevy) [inline app/services/pages/loader.rb:4]: why does the smart proxy is hard coded here? is it assumed that each page will be hard coded here or rather via an api?

Comment (ohadlevy) [inline app/services/pages/loader.rb:4]: or maybe I'm reading it wrong, and this is how the proxy is adding itself?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -220,6 +220,11 @@ def render_error(msg, status)
     render_403
   end
 
+  def render_extensible(page_view = action_name)
+    @page = ::Pages::Manager.find_page(controller_name, action_name)
+    @page.view ? render(@page.view) : fail("page #{@page.name} has no view assigned")
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb]: should this string be extract?
imho we should raise a foreman exception ?

File: app/services/pages/loader.rb (hunk 3 above)
Comment (lzap) [inline app/services/pages/loader.rb:4]: This is not extension but registration of existing pages and tabs. It is executed in the main app Rails initializer.

Plugins do have `extend*` methods to hook up on these. Similar approach can be seen in Dashboard Widgets.

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (lzap) [inline app/controllers/application_controller.rb]: Exception seems to be better, yes. Also page can be nil, this should be handled correctly.

File: app/services/pages/view_item.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        @tabs[tab_name]
+      elsif @tabs.values.empty?
+        nil
+      else
+        @tabs.values.map { |item| item.find_tab tab_name }.compact.first
+      end
+    end
+
+    def add_tab(opts)
+      fail "Cannot add tab with no name" unless opts[:name]
```
Comment (ohadlevy) [inline app/services/pages/view_item.rb]: foreman exception?

File: app/views/smart_proxies/tabs/_overview.html.erb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<% tab.columns.each do |column| %>
+  <div class="col-md-<%= 12 / tab.columns.count %>">
+    <% column.widgets.each do |widget| %>
+      <%= render widget.partial, { :smart_proxy => smart_proxy, :widget => widget } %>
+    <% end %>
+  </div>
+<% end %>
```
Comment (lzap) [inline app/views/smart_proxies/tabs/_overview.html.erb:7]: What is the benefit of having extra attribute called column here? I mean, the extended template could contain the CSS class directly, therefore there was no need of maintaining columns. It makes sense for Dashboard Widgets where things are moveable/resizable.

File: app/services/pages/loader.rb (hunk 3 above)
Comment (xprazak2) [inline app/services/pages/loader.rb:4]: `Pages::Manager` keeps track of pages that may be extended, here the page for smart_proxy/show is registered for extension

File: app/services/pages/page.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+module Pages
+  class Page < ViewItem
+    attr_reader :tabs, :widgets, :url_hash, :view
```
Comment (lzap) [inline app/services/pages/page.rb]: There are two extra attributes, leftover?

File: app/services/pages/column.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+    def find_widget(widget_name)
+      @widgets.select { |w| w.name == widget_name.to_sym }.first
+    end
+
+    def <<(widget)
+      @widgets << widget
+    end
+  end
+end
```
Comment (lzap) [inline app/services/pages/column.rb:17]: Now I see that widgets are being put into Column instances. Does this mean only table-oriented layouts are supported?

I'd expect either free-form layout where each snippet adds its own `col-x-y` class or row-oriented which better follows Bootstrap Grid pattern.

File: app/services/pages/page.rb (hunk 7 above)
Comment (xprazak2) [inline app/services/pages/page.rb]: Yes, leftovers, will remove them

**Comment by lzap** [comment]
The new commit now works on my system, thanks. -Now, what's important I think is to refactor the current Services tab into Widgets. Instead iterating over proxy features, create an extension for each particular one. Also format the page into multiple columns, I was trying this with two columns and it looks nice.-

For the record, I sent an e-mail to the dev list, here are some initial suggestions for alternatives:
- widget (like in Dashboard) 
- partial (like in Rails) 
- snippet (like in Provisioning Templates) 
- fragment (like in Rails Cache) 
- shred 
- bit 
- portion 
- portlet 
- mashup

https://groups.google.com/forum/#!topic/foreman-dev/SB1K-MX0SfI

**Comment by lzap** [comment]
I take back the refactor of current Services tab, there's no point of doing that other than trying the concept. I will test on my own plugin now.

File: app/services/pages/column.rb (hunk 8 above)
Comment (xprazak2) [inline app/services/pages/column.rb:17]: I think that layout is basically determined by the existing page/tab you are trying to extend. 
If you allow to specify the position by `col-x-y`, you have to be prepared to handle conflicts and I tried to keep it as simple as possible for the beginning, but it can potentially be enhanced into some sort of grid system.

File: app/services/foreman/plugin.rb [hunk 9]
```diff
@@ -175,6 +175,30 @@ def delete_menu_item(menu, item)
       Menu::Manager.map(menu).delete(item)
     end
 
+    # Extends an already registered extensible page. Usage:
+    #
+    # extend_page({ :controller => :controller_name, :action => :action_name } do |page|
```
Comment (lzap) [inline app/services/foreman/plugin.rb]: Missing closing `)`.

**Comment by xprazak2** [comment]
@tbrisker, more cleaned-up version of what I sent yesterday

File: app/services/pages/view_item.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      @tabs.sort_by { |k,v| v.priority }.reverse.inject({}) do |result, ary|
+        result[ary.first] = ary.last
+        result
+      end
+    end
+
+    private
+
+    def last_tab_priority
+      @tabs.values.map(&:priority).min || 20
```
Comment (lzap) [inline app/services/pages/view_item.rb:46]: Why 20? Can you also explain in the example how priorities work? All parameters could be explained there.

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -220,6 +220,15 @@ def render_error(msg, status)
     render_403
   end
 
+  def render_extensible(page_view = action_name)
+    @page = ::Pages::Manager.find_page(controller_name, action_name)
+    if @page && @page.view
+      render(@page.view)
+    else
+      raise ::ForemanException.new(N_("Page %s has no view assigned") % @page.name)
```
Comment (lzap) [inline app/controllers/application_controller.rb]: This is not the correct form, please follow this pattern: http://projects.theforeman.org/projects/foreman/wiki/Translating#Exceptions

And get rid of all `fail` methods in the patch, use this approach for all failure states that are presented to the user. Messages that go into logs do not need to be translated.

File: app/services/pages/column.rb (hunk 8 above)
Comment (lzap) [inline app/services/pages/column.rb:17]: In both cases (page or tab) the extension point is well-defined, so there can be simple assumption that every single extension point has it's own CSS row. Meaning there are 12 columns available for consumption. Plugins can simple drop elements in it, Bootstrap framework has predefined behavior how to put elements on the page. There can be conflicts of course, but as long as plugins drop elements of decent size (let's say 6) we are fine.

Do you have some working examples of plugins that add:
- one widget onto a page
- four widgets (2x2 grid) onto a page
- the same for tab

Just pastebin me working `extend_page` block.

In short - I think building a framework on top of Bootstrap Grid is not necessary and too complicated. Simple ERB/HTML is enough, the important addition here are the extension points.

**Comment by lzap** [comment]
When I try to add some code onto smart_proxies/show page, nothing happens:

```
        extend_page({ :controller => :smart_proxies, :action => :show }) do |page|
          page.add_widget :name => "W1", :partial => "some/partial"
        end
```

Since the page was added into list of extensions, I am assuming this should work. Or not? This is confusing, please explain.

File: app/controllers/smart_proxies_controller.rb [hunk 12]
```diff
@@ -8,6 +8,7 @@ def index
   end
 
   def show
+    render_extensible
```
Comment (lzap) [inline app/controllers/smart_proxies_controller.rb]: I don't understand this rendering approach. Why you override the defualt render method? Isn't better just to add before_filter which simply finds the `@page` class variable so it is available in the template? That looks cleaner to me, there might be hidden dragon you want to explain here.

**Comment by xprazak2** [comment]
> When I try to add some code onto smart_proxies/show page, nothing happens

That is because you are adding your widget directly to page and view in this case renders only widgets that are within tabs. You can add:

``` ERB
<% @page.columns.first.widgets.each do |w| %>
  <%= render w.partial %>
<% end %>
```

to the line 4 of show view, then it will appear above the tabs, or you can add it to the overview tab like this:

``` Ruby
extend_tab({:controller => smart_proxies, :action => show}, :overview) do |tab|
 tab.add_widget :name => "W1", :partial => "some/partial"
end
```

File: app/controllers/smart_proxies_controller.rb [hunk 13]
```diff
@@ -2,6 +2,7 @@ class SmartProxiesController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
 
   before_filter :find_resource, :only => [:show, :edit, :update, :refresh, :ping, :tftp_server, :destroy]
+  before_filter :find_page, :only => [:show]
```
Comment (lzap) [inline app/controllers/smart_proxies_controller.rb:5]: As per our phone call, this (and above) could be a controller concern.

File: app/services/foreman/plugin.rb [hunk 14]
```diff
@@ -175,6 +175,32 @@ def delete_menu_item(menu, item)
...
+    end
+
+    # Extends already registered tab on extensible page. Usage:
+    #
+    # extend_tab({ :controller => :controller_name, :action => :action_name }, tab_name) do |tab|
+    #  tab.add_widget(...)
+    # end
+    def extend_tab(url_hash, tab_name, &block)
+      Pages::Manager.extend_tab(url_hash, tab_name, &block)
+    end
```
Comment (lzap) [inline app/services/foreman/plugin.rb:202]: As per our phone call, I'd use more generic name than tab, because this can be used in other parts of Foreman to extend other things. Term `extend_list` comes to my mind, but I am sure you'll find a better one.

File: app/views/smart_proxies/show.html.erb [hunk 15]
```diff
@@ -1,73 +1,25 @@
...
-              <%= render :partial => "smart_proxies/plugins/no_template", :locals => {:feature => feature} %>
-            <% end %>
-          <% end %>
-        </div>
+    <% @page.sorted_tabs.each_with_index do |(name, tab), index| %>
+      <% next if name == 'services' && service_features.none? %>
+      <div class="tab-pane <%= 'active' if index == 0 %>" id="<%= tab.snake_name %>">
+        <%= render tab.layout, { :tab => tab, :smart_proxy => @smart_proxy, :service_features => service_features } %>
       </div>
     <% end %>
```
Comment (lzap) [inline app/views/smart_proxies/show.html.erb:22]: These blocks could be helper fragments or methods, so developers can simply create a container and use them.

**Comment by lzap** [comment]
I like Bryan's idea of "pagelet", seems to be a good fit.

File: app/controllers/smart_proxies_controller.rb (hunk 13 above)
Comment (lzap) [inline app/controllers/smart_proxies_controller.rb:5]: Or I was wondering if we can't simply add the before_filter to all UI controllers and not fail when there is no page/pagelet found. We can leave it set to `nil` and in the ERB we can just issue a WARNING into logs that extension point was expecting a page/pagelet definition.

The only drawback is we are adding one hash lookup for each UI request but that's not big thing.

Comment (lzap) [inline app/controllers/smart_proxies_controller.rb:5]: Or you can keep it "as is", it's a good solution as well.

File: app/services/pages/widget.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+module Pages
+  class Widget
+    attr_reader :name, :partial
+
+    def initialize(opts)
+      opts[:name] ? @name = opts[:name] : fail("Widget should have a name")
```
Comment (ohadlevy) [inline app/services/pages/widget.rb:6]: should not use fail.

**Comment by stbenjam** [comment]
How about also registering title action buttons? I tried to do that over here https://github.com/theforeman/foreman/pull/3032, but it's too simplistic.

**Comment by xprazak2** [comment]
@stbenjam, we decided to go with a bit different approach and we can extend title action buttons as well, take a look [here](https://github.com/theforeman/foreman/pull/3025)

**Comment by xprazak2** [comment]
I will close this since we will not be merging it.
